### PR TITLE
fix(models): honor TE linear backend and native initialization

### PR DESCRIPTION
**Pull request by jquesnelle**
# What does this PR do?

Honor backend.linear="te" for standard projections left by custom-model constructors while preserving native model initialization and parameter identity.

# Changelog

- Extend the existing backend override pass to replace exact nn.Linear modules when the TE linear backend is requested.
- Introduce a TE-backed, nn.Linear-compatible subtype so model-owned initialization continues to recognize the live projection.
- Preserve parameter objects, shared/tied weights, aliases, frozen parameters, training mode, and initialization flags.
- Retain existing TE backend patches and dynamically dispatch to TE’s forward implementation.
- Add CPU-compatible initialization-contract tests and real-TE forward/backward tests.

Before your PR is "Ready for review"

Pre checks:

- [x] Make sure you read and followed Contributor guidelines
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation? The compatibility subtype documents its inheritance and execution contract.

# Additional Information

## Motivation

During Qwen3-style dense pretraining, requesting backend.linear="te" left Hugging Face-derived projections as ordinary nn.Linear modules. The generic backend override pass handled Quack but not TE. Consequently, enabling TE quantization could appear successful while leaving those matrix multiplications unquantized.

Replacing modules alone is insufficient for checkpoint-free training. TE Linear is not normally an nn.Linear subclass, so model initialization based on `isinstance(module, nn.Linear)` can skip the replacement after meta-device materialization.

The fix preserves native initialization dispatch rather than imposing a generic Gaussian initializer or wrapping model owners in closures. This matters for nested configurations, identity-sensitive initialization such as module is `self.score`, and models copied before materialization for pipeline construction.

Specialized Linear subclasses and existing native TE modules remain model-owned; this conversion targets exact `nn.Linear` instances.

## Validation

Executed through SLURM with PyTorch 2.10.0+cu128 and TE 2.15.0:

- Final model-init unit and TE functional suites: 92 passed.
- Coverage includes nested and non-Gaussian initialization policies, zero initialization scales, bias/biasless projections, shared/frozen parameters, initialization flags, and deepcopy before materialization.
- Identity-sensitive BF16 head initialization is tested.
- Real-TE tests cover Qwen3 BF16/MXFP8 forward/backward and parity against native TE.
- Scoped Ruff lint/format and whitespace checks passed.

## Test command:

    python -m pytest tests/unit_tests/_transformers/test_model_init.py tests/functional_tests/training/test_te_linear_backend_init.py -q

## Broader test disclosure

The broader Transformers/Qwen3/TE test selection produced 749 passed, 1 failed, 2 skipped. The failure was:

    TestApplyPreloadOverridesPacked::test_packed_non_flash_falls_back_to_available_flash

It expected flash_attention_2 but received sdpa. The same failure reproduced on the pre-fix base, so it is not introduced by this change.

## Limitations

Validation does not establish numerical parity for every custom architecture or a complete multi-rank training matrix. No performance claims or cluster-specific dependency overrides are included.

### CODE REVIEW

File: nemo_automodel/components/models/common/te_linear.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        Args:
+            inp: Tensor of shape [..., in_features], with arbitrary leading dimensions.
+            is_first_microbatch: Whether to refresh the cached FP8 weights.
+
+        Returns:
+            Tensor of shape [..., out_features], preserving the leading dimensions.
+        """
+        return TransformerEngineLinear.forward(self, inp, is_first_microbatch=is_first_microbatch)
+
+    reset_parameters = TransformerEngineLinear.reset_parameters
```
Comment (HuiyingLi) [inline nemo_automodel/components/models/common/te_linear.py:64]: Since this class is also an `nn.Linear`, model-owned initialization may call `module.reset_parameters()`. This currently uses TE’s default initializer rather than `nn.Linear`’s Kaiming initializer. Full-checkpoint SFT is unaffected because those values are overwritten, but checkpoint-free or missing-weight paths may not preserve the original initialization contract. Could we preserve the no-argument `nn.Linear.reset_parameters()` behavior, or document this limitation?
